sdk/attestation/azure-security-attestation/samples/attestation/attest_sgxenclave_with_runtime_json.cpp

```diff
@@ -44,19 +44,19 @@ int main()
 
     // create client
     std::string const endpoint(GetEnvHelper::GetEnv("ATTESTATION_AAD_URL"));
-    AttestationClient const attestationClient(AttestationClient::Create(endpoint));
+    std::unique_ptr<AttestationClient const> attestationClient(AttestationClientFactory::Create(endpoint));
 
     std::vector<uint8_t> const sgxEnclaveQuote = AttestationCollateral::SgxQuote();
 
     // Set the RunTimeData in the request to the service. Ask the service to interpret the
     // RunTimeData as a JSON object when it is returned in the resulting token.
-    AttestEnclaveOptions attestOptions;
+    AttestSgxEnclaveOptions attestOptions;
 
     attestOptions.RunTimeData
         = AttestationData{AttestationCollateral::RunTimeData(), AttestationDataType::Json};
 
     Azure::Response<AttestationToken<AttestationResult>> const sgxResult
-        = attestationClient.AttestSgxEnclave(sgxEnclaveQuote, attestOptions);
+        = attestationClient->AttestSgxEnclave(sgxEnclaveQuote, attestOptions);
```

Comment:
I might be missing something that requires us to use this pattern, possibly specific to attestation, but we have taught users about the `client(dot)operation` experience/calling convention. As in, the client instances they use are value types and not pointers.

What's motivating us to enable and showcase this pattern to users instead for our factory create APIs?

Reply:
Anton felt STRONGLY that the factory pattern should always return a pointer and never a concrete class.

And I can understand his motivation - it adds significant flexibility w.r.t. how we express factory created classes to the customer. 

It also dramatically simplifies the factory pattern (you only have a single `Create` and not a separate `CreatePointer` method).

Ultimately, I didn't feel strong enough about the ability to create concrete class instances to insist on the ability to create concrete class instances.

Reply:
Btw, returning a `std::unique_ptr` from the factory is consistent with [C.50](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-factory).

Reply:
I am still not sure about the usability of the clients returned from the factory pattern, in how we are defining it here. I want to understand the flexibility benefit for the customer better, to add if it's worth the change in coding pattern.

@JeffreyRichter what do you think?

Here's the alternative pattern based on what the end user writes:
https://github.com/Azure/azure-sdk-for-cpp/blob/2aae6be7a36de2135a11f8c2552eb2d4143f0536/sdk/storage/azure-storage-queues/samples/queue_getting_started.cpp#L34-L39

Reply:
We *must* have the ability to create reference values in a factory function - our test framework depends on it.

@antkmsft feels strongly that factories MUST return pointers to objects and MUST NOT return concrete objects.
@ahsonkhan feels strongly that factories MUST have the ability to return concrete objects.

@ahsonkhan, I think you and @antkmsft need to get together and duke this out.  I don't have a strong preference between the two.